Mathlib/Data/Sym/Sym2.lean

```diff
@@ -100,101 +99,124 @@ It is equivalent in a natural way to multisets of cardinality 2 (see
 `Sym2.equivMultiset`).
 -/
 @[reducible]
-def Sym2 (α : Type u) :=
-  Quotient (Sym2.Rel.setoid α)
+def Sym2 (α : Type u) := Quot (Sym2.Rel α)
```

Comment:
I don't understand what the gain of moving from `Quotient` to `Quot` is here; it looks like it just makes less API available...

(I agree with the move of de-instancing the `Setoid` instance and the new notation)

Reply:
I strongly believe that any type defined using a quotient needs to recapitulate part of the quotient API so that it can ensure that there is no API leak.

The big problem with using the `Quotient` or `Quot` APIs on a type that happens to be defined using them is that (1) they change the type of the term from (for example) `Sym2` to `Quotient`, (2) they use the constructor `Quotient.mk` instead of `Sym2.mk` (again, changing the type), (3) they use non-informative `Setoid.r` for the relation rather than `Sym2.Rel`. With that last point, it's hard writing these `Quotient` relation terms yourself when there's no global Setoid instance.

I am not sure what the `Quotient` API gives us that the `Quot` API couldn't -- the main one is `Quotient.eq`, but in the long run I'm imagining quotients would be more convenient if `Quot` had lemmas that could take a class version of `Equivalence`. That way relations wouldn't unnecessarily get turned into their `Setoid`-bundled forms.

One other reason I switched from `Quotient` to `Quot` is to disable the `⟦(x, y)⟧` notation.

You'll notice that in practice we can still use `Quotient` API by passing in a setoid. There's even a place where I temporarily enable the `Setoid` instance to use it. Other places I switched to `Quot` where it was more convenient, and I added `Sym2` versions of recursors just to keep the types nice. (There's also the small point that the `Quot` recursors are missing `@[elab_as_elim]`, which needs to be fixed in core.)